File: Java/src/main/java/org/onlineoptimisation/problem/benchmark/LinearSlope.java
package org.onlineoptimisation.problem.benchmark;

import org.armadillojava.Arma;
import org.armadillojava.Col;
import org.armadillojava.Op;

public class LinearSlope extends BenchmarkProblem {

  public LinearSlope(int numberOfDimensions) {
    super(numberOfDimensions);

    if(Arma.as_scalar(Arma.randu(1, 1)) < 0.5) {
      setParameterShift(Arma.zeros(Col.class, _numberOfDimensions).plus(5));
    } else {
      setParameterShift(Arma.zeros(Col.class, _numberOfDimensions).minus(5));
    }
  }

  @Override
  public double getObjectiveValueImplementation(Col parameter) {
    Col z = new Col(parameter);
    Col outOfBound = Arma.find((z.elemTimes(_parameterShift)).strictGreaterThan(25));
    z.elem(outOfBound, Op.EQUAL, _parameterShift.elem(outOfBound));

    Col scaling = scaling(1).elemTimes(Arma.sign(_parameterShift));
    return 5 * Arma.accu(Arma.abs(scaling)) - Arma.dot(scaling, z);
  }

}

File: src/optimisationProblem/benchmark/linearSlope.cpp
#include <hop_bits/optimisationProblem/benchmark/linearSlope.hpp>

namespace hop {
  LinearSlope::LinearSlope(const unsigned int &numberOfDimensions) : BenchmarkProblem(numberOfDimensions) {

  }

  void LinearSlope::setOne(const arma::Col<double>& one) {
    _one = one;
    _xOpt = 5.0 * _one;
    _scaling = arma::sign(_one) % getScaling(10);
    _partiallyObjectiveValue = 5.0 * arma::accu(arma::abs(_scaling));
  }

  double LinearSlope::getObjectiveValueImplementation(const arma::Col<double> &parameter) const {
    arma::Col<double> z(parameter);

    arma::Col<arma::uword> outOfBound = arma::find(_xOpt % z > 25.0);
    z.elem(outOfBound) %= _xOpt.elem(outOfBound);

    return _partiallyObjectiveValue - arma::dot(_scaling, z);
  }
}
